SpatialGDK/Source/SpatialGDK/Private/EngineClasses/Components/SpatialPingComponent.cpp

```diff
@@ -0,0 +1,181 @@
+	{
+		LastSentPingID = 0;
+		TimeoutCount = 0;
+		// Set looping timer to 'tick' this component with frequency matching minimum ping interval.
+		World->GetTimerManager().SetTimer(PingTickHandle, this, &USpatialPingComponent::TickPingComponent, MinPingInterval, true);
+		bIsPingEnabled = true;
+	}
+}
+
+void USpatialPingComponent::DisablePing()
+{
+	UWorld* World = GetWorld();
+	if (World != nullptr)
+	{
+		bIsPingEnabled = false;
+		// Clear the timers.
+		World->GetTimerManager().ClearTimer(PingTickHandle);
+		World->GetTimerManager().ClearTimer(PingTimerHandle);
+		// Reset ping output value.
+		RoundTripPing = 0.f;
+	}
+}
+
+void USpatialPingComponent::TickPingComponent()
+{
+	SendNewPing();
+	// Pass latest measured ping to owning controller to be processed by PlayerState.
+	OwningController->UpdatePing(RoundTripPing);
+}
+
+void USpatialPingComponent::SendNewPing()
+{
+	// Generate a new ping ID.
+	uint16 NewPingID = LastSentPingID + 1;
+	// Send and record new ping ID.
+	SendServerWorkerPingID(NewPingID);
+	LastSentPingID = NewPingID;
+	LastSentPingTimestamp = FPlatformTime::Seconds();
+	UWorld* World = GetWorld();
+	if (World != nullptr)
+	{
+		// Set a timeout timer to await a reply.
+		World->GetTimerManager().SetTimer(PingTimerHandle, this, &USpatialPingComponent::OnPingTimeout, TimeoutLimit, false);
+	}
+}
+
+void USpatialPingComponent::OnPingTimeout()
+{
+	UWorld* World = GetWorld();
+	if (World != nullptr)
+	{
+		// Clear the timeout timer.
+		World->GetTimerManager().ClearTimer(PingTimerHandle);
+	}
+	// Update round trip ping to reflect timeout.
+	TimeoutCount++;
+	RoundTripPing = TimeoutLimit * TimeoutCount;
+	// Attempt to send another ping.
+	SendNewPing();
+}
+
+void USpatialPingComponent::OnRep_ReplicatedPingID()
+{
+	// Check that replicated ping ID matches the last sent AND check ping is enabled to catch cases where a value is replicated after being locally disabled.
+	if (ReplicatedPingID == LastSentPingID && bIsPingEnabled)
+	{
+		UWorld* World = GetWorld();
+		if (World != nullptr)
+		{
+			// Clear the timeout timer.
+			World->GetTimerManager().ClearTimer(PingTimerHandle);
+			TimeoutCount = 0;
+		}
+		// Calculate the round trip ping
+		RoundTripPing = static_cast<float>(FPlatformTime::Seconds() - LastSentPingTimestamp);
+		if (RoundTripPing >= MinPingInterval)
+		{
+			// If the current ping exceeds the min interval then send a new ping immediately.
+			SendNewPing();
+		}
+		else
+		{
+			// Otherwise set a timer for the remaining interval before sending another.
+			float RemainingInterval = MinPingInterval - RoundTripPing;
+			if (World != nullptr)
+			{
+				World->GetTimerManager().SetTimer(PingTimerHandle, this, &USpatialPingComponent::SendNewPing, RemainingInterval, false);
+			}
+		}
+	}
+}
+
+bool USpatialPingComponent::SendServerWorkerPingID_Validate(uint16 PingID)
+{
+	return true;
+}
+
+void USpatialPingComponent::SendServerWorkerPingID_Implementation(uint16 PingID)
+{
+	ReplicatedPingID = PingID;
```

Comment:
We may need to force the owning actor to replicate, otherwise the ping values we get could vary by up to several frames if this value happens to get updated just after the actor replicated.
```suggestion
	ReplicatedPingID = PingID;
	GetOwner()->ForceNetUpdate();
```

Reply:
Good call, I wasn't aware we could force it like that.